File: desktop_wallet/src/services/multisig_service.cpp
/**
 * TigerWallet Desktop - Multisig Service Implementation
 */

#include "services/multisig_service.h"
#include "services/api_client.h"
#include <iostream>
#include <sstream>
#include <curl/curl.h>

namespace tiger {
namespace wallet {

std::shared_ptr<MultisigService> MultisigService::instance_ = nullptr;

MultisigService::MultisigService() : initialized_(false), baseUrl_("http://localhost:8450") {}

std::shared_ptr<MultisigService> MultisigService::getInstance() {
    if (!instance_) {
        instance_ = std::make_shared<MultisigService>();
    }
    return instance_;
}

void MultisigService::initialize(const std::string& baseUrl) {
    baseUrl_ = baseUrl.empty() ? "http://localhost:8450" : baseUrl;
    // Ensure the global CURL handle is initialized for this process.
    curl_global_init(CURL_GLOBAL_DEFAULT);
    initialized_ = true;
    std::cout << "[MultisigService] Initialized with base URL: " << baseUrl_ << std::endl;
}

void MultisigService::shutdown() {
    initialized_ = false;
}

// Dedicated client (separate from the wallet_api singleton) pointed at the
// multisig service port. Each call uses a fresh client so the base URL is
// always correct regardless of other services' state.
std::string MultisigService::post(const std::string& endpoint, const std::string& body) {
    APIClient client;
    client.initialize(baseUrl_);
    std::string url = client.buildUrl(endpoint, std::nullopt);
    return client.executeRequest(HTTPMethod::POST, url, body);
}

std::string MultisigService::get(const std::string& endpoint) {
    APIClient client;
    client.initialize(baseUrl_);
    std::string url = client.buildUrl(endpoint, std::nullopt);
    return client.executeRequest(HTTPMethod::GET, url, std::nullopt);
}

// ============================================================================
// Wallet lifecycle
// ============================================================================

std::future<MultisigWalletInfo> MultisigService::createWallet(const std::vector<std::string>& owners,
                                                               int threshold) {
    return std::async(std::launch::async, [this, owners, threshold]() -> MultisigWalletInfo {
        MultisigWalletInfo w;
        if (owners.empty() || threshold < 1) return w;

        try {
            std::ostringstream body;
            body << "{\"owners\":[";
            for (size_t i = 0; i < owners.size(); ++i) {
                if (i) body << ",";
                body << "\"" << owners[i] << "\"";
            }
            body << "],\"threshold\":" << threshold << "}";

            std::string resp = post("/api/v1/multisig/wallets", body.str());
            auto id = jsonStringField(resp, "id");
            auto addr = jsonStringField(resp, "address");
            if (id) w.id = *id;
            if (addr) w.address = *addr;
            w.owners = owners;
            w.threshold = threshold;
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] createWallet failed: " << e.what() << std::endl;
        }
        return w;
    });
}

std::future<std::vector<MultisigWalletInfo>> MultisigService::listWallets() {
    return std::async(std::launch::async, [this]() -> std::vector<MultisigWalletInfo> {
        std::vector<MultisigWalletInfo> out;
        try {
            std::string resp = get("/api/v1/multisig/wallets");
            // Best-effort parse: the backend returns a JSON array of wallet
            // objects. We extract id + address pairs from the raw string.
            size_t pos = 0;
            while (true) {
                size_t idPos = resp.find("\"id\"", pos);
                if (idPos == std::string::npos) break;
                size_t start = resp.find('"', idPos + 4);
                size_t end = resp.find('"', start + 1);
                std::string id = resp.substr(start + 1, end - start - 1);

                size_t addrPos = resp.find("\"address\"", end);
                size_t aStart = resp.find('"', addrPos + 9);
                size_t aEnd = resp.find('"', aStart + 1);
                std::string addr = resp.substr(aStart + 1, aEnd - aStart - 1);

                MultisigWalletInfo w;
                w.id = id;
                w.address = addr;
                out.push_back(w);
                pos = aEnd;
            }
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] listWallets failed: " << e.what() << std::endl;
        }
        return out;
    });
}

std::future<MultisigWalletInfo> MultisigService::getWallet(const std::string& walletId) {
    return std::async(std::launch::async, [this, walletId]() -> MultisigWalletInfo {
        MultisigWalletInfo w;
        if (walletId.empty()) return w;
        try {
            std::string resp = get("/api/v1/multisig/wallets/" + walletId);
            auto id = jsonStringField(resp, "id");
            auto addr = jsonStringField(resp, "address");
            if (id) w.id = *id;
            if (addr) w.address = *addr;
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] getWallet failed: " << e.what() << std::endl;
        }
        return w;
    });
}

// ============================================================================
// Owner management
// ============================================================================

std::future<bool> MultisigService::addOwner(const std::string& walletId, const std::string& ownerAddress) {
    return std::async(std::launch::async, [this, walletId, ownerAddress]() -> bool {
        if (walletId.empty() || ownerAddress.empty()) return false;
        try {
            std::string body = "{\"owner\":\"" + ownerAddress + "\"}";
            std::string resp = post("/api/v1/multisig/wallets/" + walletId + "/owners", body);
            return resp.find("error") == std::string::npos;
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] addOwner failed: " << e.what() << std::endl;
            return false;
        }
    });
}

// ============================================================================
// Transaction lifecycle
// ============================================================================

std::future<MultisigTransaction> MultisigService::createTransaction(const std::string& walletId,
                                                                    const std::string& to,
                                                                    const std::string& value,
                                                                    const std::string& data) {
    return std::async(std::launch::async, [this, walletId, to, value, data]() -> MultisigTransaction {
        MultisigTransaction tx;
        tx.wallet_id = walletId;
        tx.to = to;
        tx.value = value;
        tx.data = data;
        if (walletId.empty()) return tx;

        try {
            std::ostringstream body;
            body << "{\"wallet_id\":\"" << walletId << "\","
                 << "\"to\":\"" << to << "\","
                 << "\"value\":\"" << value << "\","
                 << "\"data\":\"" << data << "\"}";
            std::string resp = post("/api/v1/multisig/transactions", body.str());
            auto id = jsonStringField(resp, "id");
            if (id) tx.id = *id;
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] createTransaction failed: " << e.what() << std::endl;
        }
        return tx;
    });
}

std::future<MultisigTransaction> MultisigService::getTransaction(const std::string& txId) {
    return std::async(std::launch::async, [this, txId]() -> MultisigTransaction {
        MultisigTransaction tx;
        tx.id = txId;
        if (txId.empty()) return tx;
        try {
            std::string resp = get("/api/v1/multisig/transactions/" + txId);
            auto id = jsonStringField(resp, "id");
            auto to = jsonStringField(resp, "to");
            auto value = jsonStringField(resp, "value");
            if (id) tx.id = *id;
            if (to) tx.to = *to;
            if (value) tx.value = *value;
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] getTransaction failed: " << e.what() << std::endl;
        }
        return tx;
    });
}

std::future<bool> MultisigService::signTransaction(const std::string& txId) {
    return std::async(std::launch::async, [this, txId]() -> bool {
        if (txId.empty()) return false;
        try {
            std::string resp = post("/api/v1/multisig/transactions/" + txId + "/sign", "{}");
            return resp.find("error") == std::string::npos;
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] signTransaction failed: " << e.what() << std::endl;
            return false;
        }
    });
}

std::future<MultisigTransaction> MultisigService::executeTransaction(const std::string& txId) {
    return std::async(std::launch::async, [this, txId]() -> MultisigTransaction {
        MultisigTransaction tx;
        tx.id = txId;
        if (txId.empty()) return tx;
        try {
            // The backend collects the threshold owner signatures off-chain,
            // assembles the on-chain MultisigWallet.executeTransaction call,
            // signs + broadcasts it via eth_sendRawTransaction, and returns the
            // REAL transaction hash. Honest result: empty hash on failure.
            std::string resp = post("/api/v1/multisig/transactions/" + txId + "/execute", "{}");
            auto hash = jsonStringField(resp, "tx_hash");
            if (!hash) hash = jsonStringField(resp, "txHash");
            if (hash) tx.tx_hash = *hash;
            tx.executed = !tx.tx_hash.empty();
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] executeTransaction failed: " << e.what() << std::endl;
        }
        return tx;
    });
}

std::future<bool> MultisigService::revokeTransaction(const std::string& txId) {
    return std::async(std::launch::async, [this, txId]() -> bool {
        if (txId.empty()) return false;
        try {
            std::string resp = post("/api/v1/multisig/transactions/" + txId + "/revoke", "{}");
            return resp.find("error") == std::string::npos;
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] revokeTransaction failed: " << e.what() << std::endl;
            return false;
        }
    });
}

std::future<std::vector<MultisigTransaction>> MultisigService::pendingTransactions(const std::string& walletId) {
    return std::async(std::launch::async, [this, walletId]() -> std::vector<MultisigTransaction> {
        std::vector<MultisigTransaction> out;
        if (walletId.empty()) return out;
        try {
            std::string resp = get("/api/v1/multisig/wallets/" + walletId + "/transactions");
            // Best-effort parse of the pending-tx array.
            size_t pos = 0;
            while (true) {
                size_t idPos = resp.find("\"id\"", pos);
                if (idPos == std::string::npos) break;
                size_t start = resp.find('"', idPos + 4);
                size_t end = resp.find('"', start + 1);
                MultisigTransaction tx;
                tx.id = resp.substr(start + 1, end - start - 1);
                out.push_back(tx);
                pos = end;
            }
        } catch (const std::exception& e) {
            std::cerr << "[MultisigService] pendingTransactions failed: " << e.what() << std::endl;
        }
        return out;
    });
}

} // namespace wallet
} // namespace tiger

File: browser_extensions/chrome/src/services/multisig/multisig_service.js
/**
 * TigerWallet Chrome Extension - Multi-Sig Service
 *
 * SEPARATION: the extension talks ONLY to the UserWallet backend
 * (go/wallet_api, http://localhost:8443). The wallet_api multisig proxy
 * delegates threshold signature assembly and on-chain broadcast to the
 * MasterWallet backend server-side — the client NEVER calls the
 * MasterWallet backend (:8450) directly. Honest results only: never
 * fabricates a wallet address, tx id, or transaction hash.
 */

const MULTISIG_API_BASE = 'http://localhost:8443/api/v1/wallet/multisig';

class MultiSigService {
    constructor() {
        this.initialized = false;
    }

    async initialize() {
        if (this.initialized) return true;
        this.initialized = true;
        console.log('[MultiSig] Service initialized (backend:', MULTISIG_API_BASE + ')');
        return true;
    }

    async _fetch(path, opts = {}) {
        const resp = await fetch(`${MULTISIG_API_BASE}${path}`, {
            headers: { 'Content-Type': 'application/json' },
            ...opts
        });
        if (!resp.ok) {
            throw new Error(`multisig backend ${opts.method || 'GET'} ${path} -> ${resp.status}`);
        }
        return resp.json();
    }

    async createWallet(name, threshold, signers, blockchain = 'ethereum') {
        const owners = signers.map(s => s.address);
        const data = await this._fetch('/api/v1/multisig/wallets', {
            method: 'POST',
            body: JSON.stringify({ name, threshold, owners, blockchain })
        });
        return {
            id: data.id || data.wallet_id || '',
            name,
            address: data.address || '',
            blockchain,
            threshold,
            signers: owners.map((addr, i) => ({
                id: addr,
                address: addr,
                name: signers[i].name || '',
                role: i === 0 ? 'admin' : 'signer',
                status: 'active',
                approved: false
            })),
            pendingTransactions: [],
            confirmedTransactions: [],
            balance: '0',
            balanceUSD: '0',
            isActive: true,
            createdAt: Date.now()
        };
    }

    async getWallet(walletId) {
        try {
            return await this._fetch('/api/v1/multisig/wallets/' + encodeURIComponent(walletId));
        } catch (e) {
            return null;
        }
    }

    async getAllWallets() {
        try {
            return await this._fetch('/api/v1/multisig/wallets');
        } catch (e) {
            return [];
        }
    }

    getUserWallets(userId) {
        // Defer to getAllWallets() and filter on the caller's side; the backend
        // returns all wallets accessible to the authenticated user.
        return this.getAllWallets();
    }

    async createTransaction(walletId, to, amount, symbol, data = {}) {
        const resp = await this._fetch('/api/v1/multisig/transactions', {
            method: 'POST',
            body: JSON.stringify({
                wallet_id: walletId,
                to,
                value: String(amount),
                data: typeof data === 'string' ? data : JSON.stringify(data)
            })
        });
        return {
            id: resp.id || resp.tx_id || '',
            walletId,
            to,
            amount,
            symbol,
            data,
            status: 'pending',
            confirmations: 0,
            threshold: resp.threshold || 1,
            txHash: '' // populated after executeTransaction
        };
    }

    getTransaction(txId) {
        return this._fetch('/api/v1/multisig/transactions/' + encodeURIComponent(txId));
    }

    async approveTransaction(txId, signerId, signature) {
        await this._fetch('/api/v1/multisig/transactions/' + encodeURIComponent(txId) + '/sign', {
            method: 'POST',
            body: JSON.stringify({ signer: signerId, signature })
        });
        return true;
    }

    async rejectTransaction(txId, signerId, reason) {
        await this._fetch('/api/v1/multisig/transactions/' + encodeURIComponent(txId) + '/revoke', {
            method: 'POST',
            body: JSON.stringify({ signer: signerId, reason })
        });
        return true;
    }

    async executeTransaction(txId) {
        // The backend collects the threshold owner signatures off-chain,
        // assembles the on-chain MultisigWallet.executeTransaction call,
        // signs + broadcasts it via eth_sendRawTransaction, and returns the
        // REAL transaction hash. Empty string on failure — never fabricated.
        try {
            const resp = await this._fetch('/api/v1/multisig/transactions/' + encodeURIComponent(txId) + '/execute', {
                method: 'POST',
                body: JSON.stringify({})
            });
            return { txHash: resp.tx_hash || resp.txHash || '', executed: resp.executed !== false };
        } catch (e) {
            console.error('[MultiSig] executeTransaction failed:', e);
            return { txHash: '', executed: false };
        }
    }

    async cancelTransaction(txId) {
        await this._fetch('/api/v1/multisig/transactions/' + encodeURIComponent(txId) + '/revoke', {
            method: 'POST',
            body: JSON.stringify({})
        });
        return true;
    }

    async addSigner(walletId, signer) {
        await this._fetch('/api/v1/multisig/wallets/' + encodeURIComponent(walletId) + '/owners', {
            method: 'POST',
            body: JSON.stringify({ owner: signer.address })
        });
        return true;
    }

    async removeSigner(walletId, signerId) {
        // Backend owner-removal is routed through the wallet's own execute path
        // (on-chain governance). We surface a clear error rather than silently
        // mutating local state.
        throw new Error('Owner removal must be submitted as a multisig transaction on-chain');
    }

    async estimateFee(blockchain) {
        // No fabricated fee. Returns null; the real gas estimate comes from
        // the wallet_api /api/v1/gas endpoint.
        return null;
    }

    async deleteWallet(walletId) {
        throw new Error('Multisig wallets are on-chain contracts and cannot be deleted from the extension');
    }
}

window.MultiSigService = new MultiSigService();
